uber/api.py

```diff
@@ -1,102 +1,217 @@
 from uber.common import *
-
-__version__ = 'v0.1'
-
-attendee_fields = [
-    'full_name', 'first_name', 'last_name', 'email', 'zip_code', 'cellphone',
-    'ec_name', 'ec_phone', 'badge_status_label', 'checked_in',
-    'badge_type_label', 'ribbon_labels', 'staffing', 'is_dept_head',
-    'assigned_depts_labels', 'weighted_hours', 'worked_hours', 'badge_num']
-
-fields = dict({
-    'food_restrictions': {
-        'sandwich_pref_labels': True,
-        'standard_labels': True,
-        'freeform': True
-    },
-    'shifts': {
-        'worked_label': True,
-        'job': [
-            'type_label', 'department_name', 'name', 'description',
-            'start_time', 'end_time', 'extra15'
-        ]
+from uber.server import register_jsonrpc
+
+
+__version__ = '0.1'
+
+
+def auth_by_token(required_access):
+    token = cherrypy.request.headers.get('X-Auth-Token', None)
+    if not token:
+        return {'error': 'Missing X-Auth-Token header'}
+
+    try:
+        token = uuid.UUID(token)
+    except ValueError as ex:
+        return {'error': 'Invalid auth token, {}: {}'.format(ex, token)}
+
+    with Session() as session:
+        api_token = session.query(ApiToken).filter_by(token=token).first()
+        if not api_token:
+            return {'error': 'Invalid auth token: {}'.format(token)}
+        if api_token.revoked_time:
+            return {'error': 'Revoked auth token: {}'.format(token)}
+        if not required_access.issubset(set(api_token.access_ints)):
+            return {'error': 'Insufficient access for auth token: {}'.format(token)}
+        cherrypy.session['account_id'] = api_token.admin_account_id
+    return None
+
+
+def auth_by_session(required_access):
+    try:
+        check_csrf()
+    except CSRFException as ex:
+        return {'error': str(ex)}
+    admin_account_id = cherrypy.session.get('account_id', None)
+    if not admin_account_id:
+        return {'error': 'Missing admin account in session'}
+    with Session() as session:
+        admin_account = session.query(AdminAccount).filter_by(id=admin_account_id).first()
+        if not admin_account:
+            return {'error': 'Invalid admin account in session'}
+        if not required_access.issubset(set(admin_account.access_ints)):
+            return {'error': 'Insufficient access for admin account'}
+    return None
+
+
+def api_auth(*required_access):
+    required_access = set(required_access)
+
+    def _decorator(func):
+        inner_func = get_innermost(func)
+        if not getattr(inner_func, 'required_access', None):
+            inner_func.required_access = required_access
+
+        @wraps(func)
+        def _with_api_auth(*args, **kwargs):
+            error = None
+            for auth in [auth_by_token, auth_by_session]:
+                result = auth(required_access)
+                error = error or result
```

Comment:
Yeah, it's not the best. I wanted to return the first error message we hit, while still checking with each available auth function. Basically, if someone sends us a bad token, I didn't want to return an error that says something like "Bad CSRF token" because it wouldn't make any sense. This bit of code can probably be reworked.